File: Python/프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
from collections import defaultdict


def solution(orders, course):
    ans = []
    
    def combination(li, n):
        result = []
        
        if n == 1:
            return li
        
        if n > 1:
            for i in range(len(li)-n+1):
                for j in combination(li[i+1:], n-1):
                    result.append(li[i] + j)
                            
        return result
        
    
    for len_c in course:
        menu_dic = defaultdict(int)
        for order in orders:
            if (len(order) < len_c): continue     
            
            for menu in combination(order, len_c):                
                menu_dic[''.join(sorted(menu))] += 1
        
        if menu_dic:
            menu_tuple = sorted(menu_dic.items(), key=lambda x:(-x[1]))
            local_max = menu_tuple[0][1]            
            if local_max < 2: continue
            
            for menu, cnt in menu_tuple:
                if cnt == local_max:
                    ans.append(menu)
    
    return sorted(ans)

File: C++/프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.cpp
#include <bits/stdc++.h>

using namespace std;

vector<string> combination(string li, int n){
    vector<string> result;
    
    if (n == 1)
        for (auto l: li)
            result.push_back(string(1, l));
    
    if (n > 1){
        for (int i=0; i < li.size()-n+1; i++){
            for (auto j: combination(li.substr(i+1), n-1)){
                result.push_back(string(1, li[i]) + j);
            }
        }   
    }
    
    return result;    
}

vector<string> solution(vector<string> orders, vector<int> course) {
    vector<string> ans;
    
    for (int num_c: course){
        unordered_map<string, int> menu_map;
        int local_max = 0;
        
        for (auto& order: orders){
            if (order.size() < num_c) continue;
            
            for (auto& menu: combination(order, num_c)){
                sort(menu.begin(), menu.end());
                menu_map[menu] += 1;
                local_max = max(local_max, menu_map[menu]);
            }            
        }
        
        if (menu_map.empty()) continue;
        if (local_max < 2) continue;
        
        for (auto& [menu, cnt]: menu_map){
            if (cnt == local_max)
                ans.push_back(menu);
        }
    }
    
    sort(ans.begin(), ans.end());
    
    return ans;
}